### PR TITLE
Memory refactoring

### PR DESCRIPTION
**Pull request by starscouts**
This basically updates the code to match the currently documented architecture.

### CODE REVIEW

**Review by ryze312 — CHANGES_REQUESTED** [review]

File: arch/src/cpu/decoder.rs [hunk 1]
```diff
@@ -17,74 +17,117 @@ pub trait Decoder {
...
-                ]) != 0 {
-                    disassembled.push_str(", ");
-                    disassembled.push_str(&Operand::disassemble(
-                        u32::from_le_bytes([
-                            opcode[0], opcode[1],
-                            opcode[2], opcode[3]
-                        ]),
-                        self, mode, disassembler_mode));
+
+        let instruction = u32::from_le_bytes([
```
[the remark this answers is not in the document]
Comment (starscouts) [reply arch/src/cpu/decoder.rs:82]: Perhaps, will try later

File: arch/src/core/instruction.rs [hunk 2]
```diff
@@ -96,3 +106,57 @@ impl Instruction {
...
+            operands,
+        })
+    }
+}
+
+impl From<DecodedInstruction<'_>> for String {
+    #[allow(clippy::cast_possible_truncation)]
+    fn from(value: DecodedInstruction) -> Self {
+        let mut disassembled = Self::new();
+        disassembled.push_str(&Operation::disassemble(value.operation as u16));
```
Comment (ryze312) [inline arch/src/core/instruction.rs:138]:
```suggestion
        let mut disassembled = Self::new();
        let operation = Operation::disassemble(value.operation as u16);
        disassembled.push_str(&operation);
```

File: arch/src/core/instruction.rs [hunk 3]
```diff
@@ -96,3 +106,57 @@ impl Instruction {
...
+                if mode != Addressing::Implied {
+                    if i > 0 {
+                        disassembled.push_str(", ");
+                    }
+                    disassembled.push_str(&Operand::disassemble(*operand, value.cpu, mode));
+                }
+            } else {
+                disassembled.push_str("<unk>");
+            }
+        }
```
Comment (ryze312) [inline arch/src/core/instruction.rs:158]: Use `.ok().filter()` and don't inline operand

```suggestion
        for (i, (operand, mode)) in operands_with_modes.iter().enumerate() {
            #[allow(clippy::cast_possible_truncation)]
            let mode = Addressing::try_from(*mode as u8)
                .ok()
                .filter(|m| m != Addressing::Implied);

            if let Some(mode) = mode {
                if i > 0 {
                    disassembled.push_str(", ");
                }
                
                let operand = Operand::disassemble(*operand, value.cpu, mode);
                disassembled.push_str(&operand);         
            } else {
                disassembled.push_str("<unk>");
            }
        }
```

**Review by ryze312 — CHANGES_REQUESTED** [review]

**Comment by ryze312** [comment]
Tried to do the second thing with an iterator, but it ended up a bit too funky, because [intersperse](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.intersperse) is unstable.

**Review by ryze312 — APPROVED** [review]
Goodie!

**Pull request merged**